src/secp256k1/src/modules/multiset/main_impl.h

```diff
@@ -0,0 +1,196 @@
+ *  Infinite uses special value, z = 0
+ */
+static void multiset_from_gej_var(secp256k1_multiset *target, const secp256k1_gej *input) {
+    if (input->infinity) {
+        memset(&target->data, 0, sizeof(target->data));
+    } else {
+        secp256k1_fe_get_b32(target->data, &input->x);
+        secp256k1_fe_get_b32(target->data+32, &input->y);
+        secp256k1_fe_get_b32(target->data+64, &input->z);
+    }
+}
+
+/** Converts a multiset to group element (Jacobian)
+ *  Infinite uses special value, z = 0
+ */
+static void gej_from_multiset_var(secp256k1_gej *target,  const secp256k1_multiset *input) {
+    secp256k1_fe_set_b32(&target->x, input->data);
+    secp256k1_fe_set_b32(&target->y, input->data+32);
+    secp256k1_fe_set_b32(&target->z, input->data+64);
+
+    target->infinity = secp256k1_fe_is_zero(&target->z) ? 1 : 0;
+}
+
+/** Converts a data element to a group element (affine)
+ *
+ *  We use trial-and-rehash which is fast but non-constant time.
+ *  Though constant time algo's exist we are not concerned with timing attacks
+ *  as we make no attempt to hide the underlying data
+ *
+ *  Pass inverse=0 to generate the group element, or inverse=1 to generate its inverse
+ */
+static void ge_from_data_var(secp256k1_ge *target, const unsigned char *input, size_t inputLen, int inverse) {
+    secp256k1_sha256 hasher;
+    unsigned char buffer[8+32];
+    unsigned char trial[32];
+    uint64_t prefix;
+
+    /* Hash to buffer, leaving space for 8-byte prefix */
+    secp256k1_sha256_initialize(&hasher);
+    secp256k1_sha256_write(&hasher, input, inputLen);
+    secp256k1_sha256_finalize(&hasher, buffer+8);
+
+    /* Loop through trials, with 50% success per loop
+     * We can assume it ends within 2^64. */
+    for(prefix=0; 1; prefix++)
+    {
+        secp256k1_fe x;
+
+        /* Set prefix in little-endian */
+        buffer[0] = prefix & 0xFF;
+        buffer[1] = (prefix>>8) & 0xFF;
+        buffer[2] = (prefix>>16) & 0xFF;
+        buffer[3] = (prefix>>24) & 0xFF;
+        buffer[4] = (prefix>>32) & 0xFF;
+        buffer[5] = (prefix>>40) & 0xFF;
+        buffer[6] = (prefix>>48) & 0xFF;
+        buffer[7] = (prefix>>56) & 0xFF;
+
+        /* Hash to trial  */
+        secp256k1_sha256_initialize(&hasher);
+        secp256k1_sha256_write(&hasher, buffer, sizeof(buffer));
+        secp256k1_sha256_finalize(&hasher, trial);
+
+        if (!secp256k1_fe_set_b32(&x, trial)) {
+            continue;
+        }
+
+        /* We let y is even be the element and odd be its inverse */
+        if (!secp256k1_ge_set_xo_var(target, &x, inverse)) {
+            continue;
+        }
+
+        VERIFY_CHECK(secp256k1_ge_is_valid_var(target));
+        VERIFY_CHECK(!secp256k1_ge_is_infinity(target));
+        break;
+    }
+}
+
+/** Adds or removes a data element */
+static int multiset_add_remove(const secp256k1_context* ctx, secp256k1_multiset *multiset, const unsigned char *input, size_t inputLen, int remove) {
```

Comment:
nit: `remove == 0` meaning it _adds_ to the multiset isn't a great interface. Can't think of a better name though.

Reply:
Hmm. This seems subjective. I am purposely "name casting"  from "remove" to "inverse" to "odd" from call to call which is as intended.